android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch

```diff
@@ -0,0 +1,1112 @@
++			return -EIO;
++		}
++	}
++	return 0;
++}
++
++
++static int bt_pcm_prepare(struct snd_pcm_substream *substream)
++{
++	struct btusb_data *data;
++	int err;
++
++	data = snd_pcm_substream_chip(substream);
++
++	// TODO: Does endpoint needs to be configured separately for Tx and Rx
++	// endpoints?
++	err = configure_endpoints(substream);
++	if (err < 0) {
++		pr_emerg("%s configure_endpoints failed err:%d\n",
++			 __func__, err);
++		return err;
++	}
++
++	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
++		pr_emerg("%s - SNDRV_PCM_STREAM_PLAYBACK\n", __func__);
++
++		/* TODO: No need to reset here */
++		clear_bit(BTUSB_ISOC_TX_START, &data->flags);
++
++		btusb_isoc_prepare_tx_urb(data);
++
++		err = btusb_isoc_initialize_tx_urb(data);
++		if (err < 0) {
++			pr_emerg("%s btusb_isoc_initialize_tx_urb failed err:%d\n",
++				 __func__, err);
++			if (data->tx_urb) {
++				usb_free_urb(data->tx_urb);
++				data->tx_urb = NULL;
++			}
++			return err;
++		}
++	} else if (substream->stream == SNDRV_PCM_STREAM_CAPTURE) {
++		pr_emerg("%s - SNDRV_PCM_STREAM_CAPTURE\n", __func__);
++
++		/* TODO: No need to reset here */
++		clear_bit(BTUSB_ISOC_RX_START, &data->flags);
++		err = btusb_isoc_prepare_rx_urbs(data);
++		if (err < 0) {
++			pr_emerg("%s btusb_isoc_prepare_rx_urb failed err:%d\n",
++				 __func__, err);
++			return err;
++		}
++	}
++	return err;
++}
++
++
++static void btusb_isoc_stop_tx(struct btusb_data *data)
++{
++	clear_bit(BTUSB_ISOC_TX_START, &data->flags);
++}
++
++static int bt_pcm_close(struct snd_pcm_substream *substream)
++{
++	struct btusb_data *data;
++
++	/* stoping the work */
++	data = snd_pcm_substream_chip(substream);
++	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
++		pr_emerg("%s SNDRV_PCM_STREAM_PLAYBACK\n", __func__);
++		clear_bit(BTUSB_ISOC_TX_START, &data->flags);
++		clear_bit(BTUSB_ISOC_TX_EP_CONFIGURED,&data->flags);
++		if (data->playback_data) {
++			kfree(data->playback_data);
++			data->playback_data = NULL;
++		}
++	} else if (substream->stream == SNDRV_PCM_STREAM_CAPTURE) {
++		pr_emerg("%s SNDRV_PCM_STREAM_CAPTURE\n", __func__);
++		clear_bit(BTUSB_ISOC_RX_START, &data->flags);
++		clear_bit(BTUSB_ISOC_RX_EP_CONFIGURED,&data->flags);
++		deinit_capture_data_cb(data->capture_data);
++	}
++
++	 if (!test_bit(BTUSB_ISOC_RX_EP_CONFIGURED, &data->flags)
++                        && !test_bit(BTUSB_ISOC_TX_EP_CONFIGURED, &data->flags)) {
++		pr_emerg("%s Reset EP and Alternate settings\n", __func__);
++
++		if (__set_isoc_interface(substream, 0) < 0){
++			printk("Resetting alt settings failed");
++			return -ENOENT;
++		}
++	}
++	return 0;
++}
++
++static int bt_pcm_hw_params(struct snd_pcm_substream *substream,
++			     struct snd_pcm_hw_params *hw_params)
++{
++	struct btusb_data *data;
++	int err;
++	unsigned long flags;
++	data = snd_pcm_substream_chip(substream);
++	err = snd_pcm_lib_malloc_pages(substream,
++			params_buffer_bytes(hw_params));
++
```

Comment:
Need to have check for supported PCM config. -> format, period size, etc checks,

Will driver work if user-space provides arbitrary period size?